src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs

```diff
@@ -14,27 +16,102 @@ namespace System.Net.Security.Tests
+                Assert.True(result, "Handshake completed in the allotted time");
+            }
+            finally
+            {
+                DisposeServerClientSslStream(clientSslStream, serverSslStream);
+                DisposeNetworkStream(clientStream, serverStream);
+            }
+        }
+
+        [Fact]
+        public void SslStream_StreamToStream_Authentication_IncorrectServerName_Fail()
+        {
             MockNetwork network = new MockNetwork();
 
             using (var clientStream = new FakeNetworkStream(false, network))
             using (var serverStream = new FakeNetworkStream(true, network))
-            using (var client = new SslStream(clientStream, false, AllowAnyServerCertificate))
+            using (var client = new SslStream(clientStream))
             using (var server = new SslStream(serverStream))
             {
-                X509Certificate2 certificate = TestConfiguration.GetServerCertificate();
                 Task[] auth = new Task[2];
-                auth[0] = client.AuthenticateAsClientAsync(certificate.Subject);
-                auth[1] = server.AuthenticateAsServerAsync(certificate);
+                auth[0] = client.AuthenticateAsClientAsync("incorrectServer");
+                auth[1] = server.AuthenticateAsServerAsync(TestConfiguration.GetServerCertificate());
+
+                Assert.Throws<AuthenticationException>(() =>
+                {
+                    auth[0].GetAwaiter().GetResult();
+                });
+
+                auth[1].GetAwaiter().GetResult();
+            }
+        }
+
+        [Fact]
+        public void SslStream_StreamToStream_Back2Back_ClientWrite_Success()
+        {
+            SslStream clientSslStream = null;
+            SslStream serverSslStream = null;
+            FakeNetworkStream clientStream = null;
+            FakeNetworkStream serverStream = null;
+            byte[] recvBuf = new byte[sampleMsg.Length];
+
+            try
+            {
+                bool result = GetServerClientSslStream(out clientSslStream, out serverSslStream, out clientStream, out serverStream, TaskWaitTime);
+
+                Assert.True(result, "Handshake completed.");
+
+                clientSslStream.Write(sampleMsg);
+
+                serverSslStream.Read(recvBuf, 0, sampleMsg.Length);
+
+                clientSslStream.Write(sampleMsg);
 
-                bool finished = Task.WaitAll(auth, TimeSpan.FromSeconds(3));
-                Assert.True(finished, "Handshake completed in the allotted time");
+                // The condition on which read method (UpdateReadStream) in FakeNetworkStream does a network read is flawed.
+                // That works fine in single read/write but fails in multi read write as stream size can be more, but real data can be < stream size.
+                // So I am doing an explicit read here. This issue is specific to test only & irrespective of xplat.
+
+                serverStream.DoNetworkRead();
```

Comment:
This isn't something that you can actually do with a NetworkStream. The comment above sounds like you're hitting a limitation of the FakeNetworkStream which should be addressed there instead of breaking encapsulation.